File: python/55_Jump_Game.py
#!/usr/bin/env python3

#易见，当出现0时不能跳转的情况才可能发生
class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """

        if nums == None or len(nums) == 0:
            return False

        if len(nums) == 1:
            return True

        if 0 not in nums:
            return True

        start = 0
        end = 0

        numsLen = len(nums)

        #特殊情况
        if(numsLen == 1):
            return 0

        while end <= numsLen - 1:
            curFarthestStep = end

            for index in range(start, end+1):
                if nums[index] + index > curFarthestStep:
                    curFarthestStep = nums[index] + index
                    if curFarthestStep >= numsLen-1:
                        return True

            flag = False
            for num in nums[end+1:curFarthestStep + 1]:#广度搜索的一轮中，如果全部是零，则终点不可达。
                if num != 0:
                    flag = True
            if not flag:
                return False

            start = end+1
            end = curFarthestStep


s = Solution()
print(s.canJump([3,2,1,0,4]))

File: java/src/brand_new/_55_Jump_Game.java
package brand_new;

public class _55_Jump_Game {
	public boolean canJump(int[] nums) {
		int p1 = 1, p2 = 0;
		for(int i = 0; i < nums.length; i++){
			if(i == p1){
				if(p2 == p1)
					return false;
				else
					p1 = p2;
			}
			
			int next = i + nums[i] + 1;
			p2 = p2 > next ? p2 : next;
		}
		return p2 >= nums.length;
	}

	public static void main(String[] args) {
		System.out.println(new _55_Jump_Game().canJump(new int[]{10,9,8,7,6,5,4,3,2,1,0,1}));
	}

}
